public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts

```diff
@@ -225,6 +230,67 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
     .addCase(rewireExpressions, (state, { payload }) => {
       state.queries = queriesWithUpdatedReferences(state.queries, payload.oldRefId, payload.newRefId);
     })
+    .addCase(removeFirstReducer, (state) => {
+      const expressionQueries = state.queries.filter((query) => isExpressionQuery(query.model));
+      const dataQueries = state.queries.filter((query) => !isExpressionQuery(query.model));
+
+      if (dataQueries.length !== 1 || expressionQueries.length !== 2) {
+        return;
+      }
+      const removedReducerQueries = state.queries.filter(
+        (query, index) =>
+          !isExpressionQuery(query.model) ||
+          (isExpressionQuery(query.model) && !(query.model.type === ExpressionQueryType.reduce && index === 1))
+      );
+      state.queries = removedReducerQueries.map((query, index) => {
+        if (index === 1 && removedReducerQueries.length === 2) {
+          // we update the only expression (threshold) to point to the query as in this case we removed the reducer
+          return {
+            ...query,
+            model: {
+              ...query.model,
+              expression: SimpleConditionIdentifier.queryId,
+            },
+          };
+        } else {
+          return query;
+        }
+      });
+    })
+    .addCase(addReducerAtFirstPosition, (state) => {
+      const expressionQueries = state.queries.filter((query) => isExpressionQuery(query.model));
+      const dataQueries = state.queries.filter((query) => !isExpressionQuery(query.model));
+      // we only add the reducer if we have one data query and one expression query. For other cases we don't do anything,
+      // and let the user add the reducer manually.
+      if (dataQueries.length !== 1 || expressionQueries.length !== 1) {
```

Comment:
I understand what you mean...I remember me wondering if we should check again this in case the consumer of the reducer is not doing this checks...because it's a precondition for the logic here.....
But I think you're right...maybe it's not necessary to re-check this here.